src/main/scala/scalaz/analytics/AnalyticsModule.scala

```diff
@@ -1,17 +1,36 @@
 package scalaz.analytics
 
 import scalaz.zio.IO
+
 import scala.language.implicitConversions
 
 /**
  * An analytics module defined the abstract components required to build a reified program description.
  */
 trait AnalyticsModule {
 
+  /**
+   * An abstract Dataset
+   */
+  type Dataset[A]
+
   /**
    * An abstract DataStream.
    */
-  type DataStream[_]
+  type DataStream[A]
+
+  /**
+   * A window is a way to specify a view on a DataStream/Dataset
+   */
+  sealed trait Window
+
+  object Window {
+    case class FixedTimeWindow()   extends Window
+    case class SlidingTimeWindow() extends Window
+    case class SessionWindow()     extends Window
+    // For internal use only
+    case class GlobalWindow private () extends Window
```

Comment:
I don't _think_ we want to expose the `GlobalWindow` to users since they will be working with `Dataset` when they want a global window. Doing so on a stream is non sensical anyway. Does anyone have any objections? We can always revisit this if we run into something down the line.

Reply:
`GlobalWindow` here represents the complete stream?

Reply:
Yeah - it essentially represents a bounded stream. Applying it to an infinite stream would be non terminating - which is why I made it private. I'm pretty sure we only need it internally for building the DSL in `AnalyticsModule` for now.

Reply:
Alright . Also just a random thought here, not sure how to the point it is but isn't `GlobalWindow` can essentially be represented as `DataSet`. If thats so then isn't the role becomes redundant ?

Reply:
Yeah so from an API perspective the concept of a `GlobalWindow` never needs to be exposed to users (hence making the constructor private). They will interact directly with `Dataset` and `DataStream`. It is purely so that we can represent bounded batch computations with the same grammar as bounded stream computations (i.e. Stream with a `GlobalWindow`). If we don't do this, then we need two versions of every bounded method (one with a window and one without) - despite them being identical semantically.

Take a look at Beam - they have a very similar concept (everything is a stream, but batches are a special type of stream): https://beam.apache.org/documentation/programming-guide/#windowing